include/devicetree.h

```diff
@@ -5,10 +5,1273 @@
+
+/**
+ * @}
+ */
+
+/**
+ * @defgroup devicetree-generic-bus Bus and hardware specific APIs
+ * @{
+ */
+
+/**
+ * @brief Node's bus controller
+ *
+ * Get the node identifier of the node's bus controller. This can be
+ * used with @ref DT_PROP() to get properties of the bus controller.
+ *
+ * It is an error to use this with nodes which do not have bus
+ * controllers.
+ *
+ * Example devicetree fragment:
+ *
+ *     i2c@deadbeef {
+ *             label = "I2C_CTLR";
+ *             status = "okay";
+ *             clock-frequency = < 100000 >;
+ *
+ *             i2c_device: accelerometer@12 {
+ *                     ...
+ *             };
+ *     };
+ *
+ * Example usage, which evaluates to 100000:
+ *
+ *     DT_PROP(DT_BUS(DT_NODELABEL(i2c_device)), clock_frequency)
+ *
+ * @param node_id node identifier
+ * @return a node identifier for the node's bus controller
+ */
+#define DT_BUS(node_id) DT_CAT(node_id, _BUS)
+
+/**
+ * @brief Node's bus controller's label property
+ * @param node_id node identifier
+ * @return the label property of the node's bus controller DT_BUS(node)
+ */
+#define DT_BUS_LABEL(node_id) DT_PROP(DT_BUS(node_id), label)
+
+/**
+ * @brief Test if a node's bus type is a given type
+ *
+ * Example devicetree overlay:
+ *
+ *     &i2c0 {
+ *            temp: temperature-sensor@76 {
+ *                     compatible = "vnd,some-sensor";
+ *                     reg = <0x76>;
+ *            };
+ *     };
+ *
+ * Example usage which evaluates to 1, assuming "i2c0" is an I2C bus
+ * controller node, and therefore "temp" is on an I2C bus:
+ *
+ *     DT_ON_BUS(DT_NODELABEL(temp), i2c)
+ *
+ * Example usage which evaluates to 0, since the "temp" node isn't on
+ * a SPI bus:
+ *
+ *     DT_ON_BUS(DT_NODELABEL(temp), spi)
+ *
+ * @param node_id node identifier
+ * @param bus a binding's bus type as a C token, lowercased and without quotes
+ * @return 1 if the node is on a bus of the given type,
+ *         0 otherwise
+ */
+#define DT_ON_BUS(node_id, bus) IS_ENABLED(DT_CAT(node_id, _BUS_##bus))
+
+/**
+ * @brief Get gpio controller "name" (label property) at an index
+ *
+ * It's an error if the GPIO controller referenced by the phandle
+ * in property "gpio_pha" at index "idx" has no label property.
+ *
+ * Example devicetree fragment:
+ *
+ *     gpio1: gpio@... {
+ *             label = "GPIO_1";
+ *     };
+ *
+ *     gpio2: gpio@... {
+ *             label = "GPIO_2";
+ *     };
+ *
+ *     n: node {
+ *             gpios = <&gpio1 10 20>, <&gpio2 30 40>;
+ *     };
+ *
+ * Example usage, which evaluates to "GPIO_2":
+ *
+ *     DT_GPIO_LABEL_IDX(DT_NODELABEL(n), gpios, 1)
+ *
+ * @param node_id node identifier
+ * @param gpio_pha lowercase-and-underscores GPIO property with
+ *        type "phandle-array"
+ * @param idx logical index into the property
+ * @return the label property for the referenced node at index idx
+ * @see DT_PHA_PHANDLE_IDX()
+ */
+#define DT_GPIO_LABEL_IDX(node_id, gpio_pha, idx) \
```

Comment:
We should pull the DT_GPIO macro's out into either into include/drivers/gpio.h or create a new devicetree header group and put this in include/devicetree/gpio.h.  The reason is we'll need similar macro's for IO_CONTROLLERs, CLOCKS, and any other phandle/phandle-array.  @mnkp / @pabigot any thoughts?

Reply:
API meeting:

subsystem-specific DT helper macros belong either in:

`include/devicetree/<foo>.h`
or
`include/drivers/<foo>.h`

If `include/devicetree.h` includes `include/devicetree/*` then the application only ever needs to include `<devicetree.h>`.

Quorum: `include/devicetree/*`